File: CSLP_work3_python/Bitstream.py
import math
import os


class Bitstream:
    def __init__(self, filename):
        self.bitstream = []
        self.filename = filename
        self.file = None
        self.padding_last_byte = 0
        self.file_size = 0
        self.bytes_read = 0

    def read_one_bit(self):  # or is it to get a bit from the bitstream????
        if not self.bitstream:
            self.read_from_file(True)
        return self.bitstream.pop(0)

    def write_one_bit(self, bit):
        assert bit == 1 or bit == 0
        if len(self.bitstream) == 8:
            self.write_to_file()
        self.bitstream.append(str(bit))

        return bit

    def read_n_bits(self, start, k):
        # print("hiiiiiiii: ", self.bitstream)
        while True:
            if len(self.bitstream) - 1 >= k:
                temp = self.bitstream[start:start + k]
                del self.bitstream[start:start + k]
                return temp
            else:
                self.read_from_file(True)

    def write_n_bits(self, r, m):
        assert r < m
        bits = []
        b = math.log2(m)
        if not b.is_integer():
            b = int(math.ceil(math.log2(m)))
            level = int(math.pow(2, b) - m)

            if r < level:
                bits = bin(r)
                bits = list(bits[2:])

                while len(bits) < b - 1:
                    bits = ['0'] + bits

            else:
                r = r + level
                bits = bin(r)
                bits = list(bits[2:])

                while len(bits) < b:
                    bits = ['0'] + bits

        else:
            b = int(math.ceil(math.log2(m)))
            level = int(math.pow(2, b) - m)

            if r < level:
                bits = bin(r)
                bits = list(bits[2:])

                while len(bits) < b - 1:
                    bits = ['0'] + bits

            else:
                r = r + level
                bits = bin(r)
                bits = list(bits[2:])

                while len(bits) < b:
                    bits = ['0'] + bits

        for x in bits:
            if len(self.bitstream) == 8:
                self.write_to_file()
            self.bitstream.append(x)

        return bits

    def write_to_file(self):
        #print("Writing to file... ", self.bitstream)
        self.file_size += 1
        value = int("".join(str(x) for x in self.bitstream), 2)
        self.bitstream.clear()
        self.file.write(bytes([value]))
        # print("Done")

    def read_from_file(self, special):
        self.bytes_read += 1
        by = self.file.read(1)
        value = int.from_bytes(by, byteorder="little")
        bits = [int(digit) for digit in bin(value)[2:]]
        result = [str(b) for b in bits]
        # print("Reading from file... ", result, self.bytes_read, self.file_size, self.bitstream)

        while len(result) < 8:
            result = ['0'] + result

        if self.bytes_read == self.file_size:
            for i in range(self.padding_last_byte):
                result.pop(0)

        self.bitstream += result
        # print("State of bitstream: ", self.bitstream)

    def open_file_write(self):
        self.file = open(self.filename, "wb")

    def open_file_read(self):
        self.file = open(self.filename, "rb")

    def close_file(self):
        self.file.close()


def main():
    bs = Bitstream("outputfile.txt")

    print(bs.write_n_bits(15, 18))
    print(bs.read_one_bit())
    print(bs.read_n_bits(1, 2))
    bs.open_file_write()
    bs.write_to_file()
    bs.close_file()
    bs.open_file_read()
    bs.read_from_file(True)
    bs.close_file()
    print(bs.bitstream)


if __name__ == "__main__":
    main()

File: CSLP_work3_Cpp/Bitstream.cpp
/*
 * Bitstream.cpp
 *
 *  Created on: 21 Nov 2019
 *      Author: ghost
 */

#include "Bitstream.h"
#include <functional>
#include <cmath>
#include <bitset>
#include <string>
#include <fstream>
#include <iostream>

using namespace std;

ifstream file;

Bitstream::Bitstream() {
	bitstream = "";
	filename = "output.bin";
	bytes = new char[1];
	readerCounter = 0;
}

void Bitstream::writeOneBit(int bit){
	if(bitstream.length() == 8){
		writeToFile();
	}

	bitstream += std::to_string(bit);

}

std::string Bitstream::getBitstream(){
	return bitstream;
}

void Bitstream::writeNBits(int r, int m){
	double b = log2(m);
	std::string binr = "";
	if(ceil(b) != floor(b)){
		int b2 = int(ceil(b));
			int level = int(pow(2, b2) - m);

			if(r < level){
				binr = std::bitset<8>(r).to_string();

				while(binr.length() > b2 - 1){
					binr.erase(binr.begin());
				}
			}

			else{
				r += level;
				binr = std::bitset<8>(r).to_string();

				while(binr.length() > b2){
					binr.erase(binr.begin());
				}
			}
	}
	else{
		//cout << "hey" << "\n";
		int b2 = int(ceil(b));
					int level = int(pow(2, b2) - m);

					if(r < level){
						binr = std::bitset<8>(r).to_string();

						while(binr.length() > b2 - 1){
							binr.erase(binr.begin());
						}
					}

					else{
						r += level;
						binr = std::bitset<8>(r).to_string();

						while(binr.length() > b2){
							binr.erase(binr.begin());
						}
					}
	}
	for(int i = 0; i < binr.length(); i++){
		if(bitstream.length() == 8){
			writeToFile();
		}
		char c = binr.at(i);
		bitstream += c;
	}

}

char Bitstream::readOneBit(){
	if(bitstream.length() == 0){
		readFromFile();
	}
	char bit = bitstream.at(0);
	bitstream.erase(bitstream.begin());
	return bit;

}

std::string Bitstream::readNBits(int start, int k){
	if(bitstream.length() - 1 < k){
		readFromFile();
	}
	//cout << "error: " << bitstream << endl;
	std::string result = bitstream.substr(start, k);
	bitstream.erase(start, k);
	//cout << "new: " << bitstream << endl;
	return result;
}

void Bitstream::readFromFile(){
	//cout << "Getting byte" << endl;
	int value = (int)bytes[readerCounter];
	if(value < 0)
		value = value + 256;
	std::string binr = std::bitset<8>(value).to_string();
	readerCounter++;
	if(readerCounter == filesize){
		int diff = 8 - lastByteLength;
		cout << "diff: " << diff << endl;
		binr = binr.substr(diff, (int)binr.length());
	}
	bitstream += binr;
	cout << "state of bitstream: " << binr << endl;

}

void Bitstream::writeToFile(){
	fstream file (filename, ios::out | ios::binary | ios::app);
	std::bitset<8> b(bitstream);
	int c = (int)b.to_ulong();
	cout << "Hey!!!" << b.to_string() << "hey: " << c << endl;
	int x[1] = {c};
	cout << x[0] << endl;
	file.write((char *)&x, 1);
	bitstream = "";
	file.close();
}

void Bitstream::readFile(){
	fstream file (filename, ios::in | ios::binary | ios::ate);
	filesize = file.tellg();
	cout << "Size: " << filesize << endl;
	bytes = new char[filesize];
	file.seekg(0, ios::beg);
	file.read(bytes, filesize);
	file.close();
	remove("output.bin");
}
/*
int main(){
	Bitstream bt;
	bt.writeOneBit(1);
	cout << "1: " << bt.getBitstream() << "\n";
	bt.writeNBits(2, 4);
	cout << "2: " << bt.getBitstream() << "\n";
	bt.writeToFile();
	bt.readFile();
	cout << "3: " << bt.readOneBit() << "\n";
	cout << "4: " << bt.readNBits(0, 2) << "\n";

}
*/



